File: python/algoforge/backtest.py
"""BacktestEngine — bar-by-bar simulation engine.

Mirrors the C++ BacktestEngine blueprint exactly:
  - ATR-based SL/TP
  - Fixed-fractional position sizing (risk_pct of capital per trade)
  - Commission in USD per lot
  - Slippage in points (1 pt = 0.00001)
  - Equity-curve-based metrics (Sharpe, Sortino, Calmar, max-drawdown …)

Loop semantics (matching C++):
  At bar i (i in [warmup_bars, n-1]):
    1. Check SL/TP on open trade using bar[i].high / bar[i].low.
    2. If no open trade, run indicators+patterns on bars[0..i-1] (i.e. the
       *previous* bar), obtain a decision, and fill at bar[i].open ± slippage.
       This ensures no look-ahead: the signal comes from completed bar i-1,
       and execution happens at the open of bar i.

Usage::

    from algoforge.backtest import BacktestEngine, BTConfig
    from algoforge.algorithm import StubAlgo

    engine = BacktestEngine(StubAlgo())
    result = engine.run(bars, symbol="EURUSD", tf=3600)
    print(result.summary())
"""

from __future__ import annotations

import math
import time
from dataclasses import dataclass, field
from typing import Optional

from .algorithm import AlgoDecision, AlgoSignal, IAlgorithm
from .indicators import EngineResult, IndicatorEngine
from .patterns import PatternEngine
from .types import Bar, Direction


# ── Configuration ─────────────────────────────────────────────────────────────

@dataclass
class BTConfig:
    """Backtest configuration — all defaults match the C++ BTConfig."""
    initial_capital: float = 10_000.0
    risk_pct:        float = 0.01          # fraction of capital risked per trade
    commission_usd:  float = 7.0           # USD per lot
    slippage_pts:    int   = 2             # points of slippage (1 pt = 0.00001)
    atr_sl_mult:     float = 1.5           # ATR multiplier for stop-loss
    rr_ratio:        float = 2.0           # reward-to-risk ratio (TP = SL * rr_ratio)
    warmup_bars:     int   = 200           # bars skipped before trading begins


# ── Trade record ──────────────────────────────────────────────────────────────

@dataclass
class BTTrade:
    """Record of a single completed trade."""
    entry_time:   int       = 0
    exit_time:    int       = 0
    symbol:       str       = ""
    direction:    Direction = Direction.NONE
    lots:         float     = 0.0
    entry_price:  float     = 0.0
    exit_price:   float     = 0.0
    sl:           float     = 0.0
    tp:           float     = 0.0
    gross_pnl:    float     = 0.0
    commission:   float     = 0.0
    bars_held:    int       = 0
    close_reason: str       = ""   # "SL" | "TP" | "END"
    algo_name:    str       = ""

    @property
    def net_pnl(self) -> float:
        return self.gross_pnl - self.commission

    @property
    def is_winner(self) -> bool:
        return self.net_pnl > 0.0


# ── Result container with computed metrics ────────────────────────────────────

@dataclass
class BTResult:
    """Backtest result — trades, equity curve, and performance metrics."""
    symbol:          str           = ""
    timeframe:       int           = 0
    initial_capital: float         = 10_000.0
    trades:          list[BTTrade] = field(default_factory=list)
    equity_curve:    list[float]   = field(default_factory=list)
    duration_sec:    float         = 0.0

    # ── Basic aggregates ──────────────────────────────────────────────────────

    @property
    def total_trades(self) -> int:
        return len(self.trades)

    @property
    def winners(self) -> list[BTTrade]:
        return [t for t in self.trades if t.is_winner]

    @property
    def losers(self) -> list[BTTrade]:
        return [t for t in self.trades if not t.is_winner]

    @property
    def net_profit(self) -> float:
        return sum(t.net_pnl for t in self.trades)

    @property
    def final_capital(self) -> float:
        return self.initial_capital + self.net_profit

    @property
    def return_pct(self) -> float:
        return self.net_profit / self.initial_capital * 100.0

    @property
    def win_rate(self) -> float:
        if not self.trades:
            return 0.0
        return len(self.winners) / len(self.trades)

    @property
    def avg_win(self) -> float:
        w = self.winners
        if not w:
            return 0.0
        return sum(t.net_pnl for t in w) / len(w)

    @property
    def avg_loss(self) -> float:
        l = self.losers
        if not l:
            return 0.0
        return sum(t.net_pnl for t in l) / len(l)

    @property
    def profit_factor(self) -> float:
        gross_win  = sum(t.net_pnl for t in self.winners)
        gross_loss = sum(abs(t.net_pnl) for t in self.losers)
        if gross_loss < 1e-12:
            return 999.0
        return gross_win / gross_loss

    @property
    def expectancy(self) -> float:
        wr = self.win_rate
        return wr * self.avg_win + (1.0 - wr) * self.avg_loss

    # ── Equity-curve metrics ──────────────────────────────────────────────────

    @property
    def max_drawdown_pct(self) -> float:
        """Peak-to-trough drawdown on equity_curve expressed as a percentage."""
        curve = self.equity_curve
        if len(curve) < 2:
            return 0.0
        peak   = curve[0]
        max_dd = 0.0
        for v in curve:
            if v > peak:
                peak = v
            dd = (peak - v) / peak * 100.0 if peak > 1e-12 else 0.0
            if dd > max_dd:
                max_dd = dd
        return max_dd

    @property
    def sharpe_ratio(self) -> float:
        """Annualised Sharpe ratio (sqrt(252) * mean/std of bar-to-bar returns)."""
        curve = self.equity_curve
        if len(curve) < 2:
            return 0.0
        returns: list[float] = []
        for i in range(1, len(curve)):
            prev = curve[i - 1]
            returns.append((curve[i] - prev) / prev if prev > 1e-12 else 0.0)
        if not returns:
            return 0.0
        mean     = sum(returns) / len(returns)
        variance = sum((r - mean) ** 2 for r in returns) / len(returns)
        std      = math.sqrt(variance)
        if std < 1e-12:
            return 0.0
        return math.sqrt(252.0) * mean / std

    @property
    def sortino_ratio(self) -> float:
        """Annualised Sortino ratio (uses only negative-return std)."""
        curve = self.equity_curve
        if len(curve) < 2:
            return 0.0
        returns: list[float] = []
        for i in range(1, len(curve)):
            prev = curve[i - 1]
            returns.append((curve[i] - prev) / prev if prev > 1e-12 else 0.0)
        if not returns:
            return 0.0
        mean         = sum(returns) / len(returns)
        neg_returns  = [r for r in returns if r < 0.0]
        if not neg_returns:
            return 999.0
        downside_var = sum(r ** 2 for r in neg_returns) / len(neg_returns)
        downside_std = math.sqrt(downside_var)
        if downside_std < 1e-12:
            return 999.0
        return math.sqrt(252.0) * mean / downside_std

    @property
    def calmar_ratio(self) -> float:
        """Return_pct divided by max_drawdown_pct."""
        mdd = self.max_drawdown_pct
        if mdd < 1e-12:
            return 999.0
        return self.return_pct / mdd

    # ── Human-readable summary ────────────────────────────────────────────────

    def summary(self) -> str:
        lines = [
            f"=== Backtest Result: {self.symbol} (tf={self.timeframe}) ===",
            f"  Duration         : {self.duration_sec:.3f}s",
            f"  Initial Capital  : ${self.initial_capital:,.2f}",
            f"  Final Capital    : ${self.final_capital:,.2f}",
            f"  Net Profit       : ${self.net_profit:,.2f}",
            f"  Return           : {self.return_pct:.2f}%",
            f"  Total Trades     : {self.total_trades}",
            f"  Win Rate         : {self.win_rate * 100:.1f}%",
            f"  Profit Factor    : {self.profit_factor:.2f}",
            f"  Avg Win          : ${self.avg_win:.2f}",
            f"  Avg Loss         : ${self.avg_loss:.2f}",
            f"  Expectancy       : ${self.expectancy:.2f}",
            f"  Max Drawdown     : {self.max_drawdown_pct:.2f}%",
            f"  Sharpe Ratio     : {self.sharpe_ratio:.4f}",
            f"  Sortino Ratio    : {self.sortino_ratio:.4f}",
            f"  Calmar Ratio     : {self.calmar_ratio:.4f}",
        ]
        return "\n".join(lines)


# ── Engine ────────────────────────────────────────────────────────────────────

_POINT: float         = 0.00001       # 1 point in price terms
_CONTRACT_SIZE: float = 100_000.0     # standard forex contract size (1 lot)


class BacktestEngine:
    """Bar-by-bar backtest engine.

    Parameters
    ----------
    algo:
        The trading algorithm to evaluate.
    cfg:
        Backtest configuration.  Defaults to :class:`BTConfig` defaults.
    """

    def __init__(self, algo: IAlgorithm, cfg: BTConfig = BTConfig()) -> None:
        self._algo = algo
        self._cfg  = cfg
        self._ind  = IndicatorEngine()
        self._pat  = PatternEngine()

    # ── Public API ────────────────────────────────────────────────────────────

    def run(self, bars: list[Bar], symbol: str, tf: int) -> BTResult:
        """Run the backtest over *bars* and return a :class:`BTResult`."""
        t0  = time.perf_counter()
        cfg = self._cfg

        result = BTResult(
            symbol          = symbol,
            timeframe       = tf,
            initial_capital = cfg.initial_capital,
        )

        n = len(bars)
        # Need at least warmup_bars + 1 so we have a bar to evaluate and a next
        # bar to fill on.
        if n < cfg.warmup_bars + 1:
            result.duration_sec = time.perf_counter() - t0
            return result

        capital:  float = cfg.initial_capital
        slippage: float = cfg.slippage_pts * _POINT

        # Open-trade state (None = flat)
        open_trade:    Optional[BTTrade] = None
        entry_bar_idx: int               = 0

        equity_curve: list[float] = [capital]

        # Loop from warmup_bars to n-1 inclusive.
        # At each bar i:
        #   - Check SL/TP using bar[i].high / bar[i].low.
        #   - If flat, evaluate signal at bar[i-1] and fill at bar[i].open.
        for i in range(cfg.warmup_bars, n):
            bar = bars[i]

            # ── Step 1: Check SL / TP on open trade ───────────────────────
            if open_trade is not None:
                closed = False

                if open_trade.direction == Direction.LONG:
                    if bar.low <= open_trade.sl:
                        # SL hit — exit at SL price
                        exit_px = open_trade.sl
                        gross   = ((exit_px - open_trade.entry_price)
                                   * open_trade.lots * _CONTRACT_SIZE)
                        self._close_trade(open_trade, bar.timestamp, exit_px,
                                          gross, "SL", i - entry_bar_idx)
                        capital += open_trade.net_pnl
                        result.trades.append(open_trade)
                        open_trade = None
                        closed = True
                    elif bar.high >= open_trade.tp:
                        # TP hit — exit at TP price
                        exit_px = open_trade.tp
                        gross   = ((exit_px - open_trade.entry_price)
                                   * open_trade.lots * _CONTRACT_SIZE)
                        self._close_trade(open_trade, bar.timestamp, exit_px,
                                          gross, "TP", i - entry_bar_idx)
                        capital += open_trade.net_pnl
                        result.trades.append(open_trade)
                        open_trade = None
                        closed = True

                elif open_trade.direction == Direction.SHORT:
                    if bar.high >= open_trade.sl:
                        # SL hit
                        exit_px = open_trade.sl
                        gross   = ((open_trade.entry_price - exit_px)
                                   * open_trade.lots * _CONTRACT_SIZE)
                        self._close_trade(open_trade, bar.timestamp, exit_px,
                                          gross, "SL", i - entry_bar_idx)
                        capital += open_trade.net_pnl
                        result.trades.append(open_trade)
                        open_trade = None
                        closed = True
                    elif bar.low <= open_trade.tp:
                        # TP hit
                        exit_px = open_trade.tp
                        gross   = ((open_trade.entry_price - exit_px)
                                   * open_trade.lots * _CONTRACT_SIZE)
                        self._close_trade(open_trade, bar.timestamp, exit_px,
                                          gross, "TP", i - entry_bar_idx)
                        capital += open_trade.net_pnl
                        result.trades.append(open_trade)
                        open_trade = None
                        closed = True

                equity_curve.append(capital)
                if closed:
                    continue

            # ── Step 2: If flat, evaluate signal on prev bar → fill this bar ─
            if open_trade is None:
                # Evaluate indicator + pattern engines on bar[i-1]
                prev_idx: int              = i - 1
                ind_result: EngineResult   = self._ind.compute(bars, prev_idx)
                decision:   AlgoDecision   = self._algo.safe_evaluate(
                    symbol, tf, bars, prev_idx, ind_result, self._pat
                )

                if decision.is_actionable() and ind_result.atr_value is not None:
                    atr_val = ind_result.atr_value
                    if atr_val > 0.0:
                        signal    = decision.signal
                        direction = (Direction.LONG
                                     if signal == AlgoSignal.BUY
                                     else Direction.SHORT)

                        if direction == Direction.LONG:
                            entry_price = bar.open + slippage
                            sl = entry_price - atr_val * cfg.atr_sl_mult
                            tp = entry_price + atr_val * cfg.atr_sl_mult * cfg.rr_ratio
                        else:  # SHORT
                            entry_price = bar.open - slippage
                            sl = entry_price + atr_val * cfg.atr_sl_mult
                            tp = entry_price - atr_val * cfg.atr_sl_mult * cfg.rr_ratio

                        stop_dist = abs(entry_price - sl)
                        if stop_dist > 1e-12:
                            # Position sizing: lots = capital * risk_pct / (stop_dist * 100000)
                            lots_raw = capital * cfg.risk_pct / (stop_dist * _CONTRACT_SIZE)
                            lots     = max(0.01, min(10.0, round(lots_raw, 2)))
                            commission = cfg.commission_usd * lots

                            open_trade = BTTrade(
                                entry_time  = bar.timestamp,
                                symbol      = symbol,
                                direction   = direction,
                                lots        = lots,
                                entry_price = entry_price,
                                sl          = sl,
                                tp          = tp,
                                commission  = commission,
                                algo_name   = self._algo.name(),
                            )
                            entry_bar_idx = i

                equity_curve.append(capital)

        # ── Step 3: Close any remaining open trade at last bar's close ────────
        if open_trade is not None:
            last_bar   = bars[-1]
            exit_price = last_bar.close
            if open_trade.direction == Direction.LONG:
                gross = ((exit_price - open_trade.entry_price)
                         * open_trade.lots * _CONTRACT_SIZE)
            else:
                gross = ((open_trade.entry_price - exit_price)
                         * open_trade.lots * _CONTRACT_SIZE)
            bars_held = n - 1 - entry_bar_idx
            self._close_trade(open_trade, last_bar.timestamp, exit_price,
                              gross, "END", bars_held)
            capital += open_trade.net_pnl
            result.trades.append(open_trade)
            # Patch the last equity point to reflect the closed trade
            if equity_curve:
                equity_curve[-1] = capital
            else:
                equity_curve.append(capital)

        result.equity_curve = equity_curve
        result.duration_sec = time.perf_counter() - t0
        return result

    # ── Internal helpers ──────────────────────────────────────────────────────

    @staticmethod
    def _close_trade(
        trade:        BTTrade,
        exit_time:    int,
        exit_price:   float,
        gross_pnl:    float,
        reason:       str,
        bars_held:    int,
    ) -> None:
        """Mutate *trade* in-place to record exit details."""
        trade.exit_time    = exit_time
        trade.exit_price   = exit_price
        trade.gross_pnl    = gross_pnl
        trade.bars_held    = bars_held
        trade.close_reason = reason

File: js/src/backtest.js
'use strict';

/**
 * BacktestEngine — mirrors C++ BacktestEngine exactly.
 *
 * Indicator and pattern functions are imported from the sibling ESM modules.
 * Only atr() is used internally; the full ind/pat objects are passed to
 * algo.safeEvaluate() so real algorithms can inspect them.
 */

import { atr as _atrFn } from './indicators.js';
import { AlgoSignal }     from './algorithm.js';

// ── Private helpers ───────────────────────────────────────────────────────────

/**
 * Build indicator arrays for the current bar window.
 * Exposes atr14 (14-period ATR) as the primary risk sizing input.
 * @param {object[]} bars
 * @returns {{ atr14: Array<number|null> }}
 */
function _buildIndicators(bars) {
    const highs  = bars.map(b => b.high);
    const lows   = bars.map(b => b.low);
    const closes = bars.map(b => b.close);
    return { atr14: _atrFn(highs, lows, closes, 14) };
}

/**
 * Build a lightweight pattern context passed to algo.evaluate().
 * Real algorithms may perform their own full scan; the engine itself
 * only needs the indicator values for SL/TP sizing.
 * @param {object[]} bars
 * @returns {{ matches: [] }}
 */
function _buildPatterns(/* bars */) {
    return { matches: [] };
}

// ── BTConfig ──────────────────────────────────────────────────────────────────

export class BTConfig {
    constructor({
        initialCapital = 10000,
        riskPct        = 0.01,
        commissionUsd  = 7.0,
        slippagePts    = 2,
        atrSlMult      = 1.5,
        rrRatio        = 2.0,
        warmupBars     = 200,
    } = {}) {
        this.initialCapital = initialCapital;
        this.riskPct        = riskPct;
        this.commissionUsd  = commissionUsd;
        this.slippagePts    = slippagePts;
        this.atrSlMult      = atrSlMult;
        this.rrRatio        = rrRatio;
        this.warmupBars     = warmupBars;
    }
}

// ── BTTrade ───────────────────────────────────────────────────────────────────

/**
 * Factory that creates a BTTrade plain object with methods attached.
 * All fields mirror C++ BTTrade exactly.
 *
 * @param {{
 *   entryTime?: number, exitTime?: number, symbol?: string,
 *   direction?: number, lots?: number, entryPrice?: number,
 *   exitPrice?: number, sl?: number, tp?: number,
 *   grossPnl?: number, commission?: number, barsHeld?: number,
 *   closeReason?: string, algoName?: string
 * }} opts
 */
export function makeBTTrade({
    entryTime   = 0,
    exitTime    = 0,
    symbol      = '',
    direction   = 0,   // 0=NONE, 1=LONG, -1=SHORT
    lots        = 0,
    entryPrice  = 0,
    exitPrice   = 0,
    sl          = 0,
    tp          = 0,
    grossPnl    = 0,
    commission  = 0,
    barsHeld    = 0,
    closeReason = '',  // "SL" | "TP" | "END"
    algoName    = '',
} = {}) {
    return {
        entryTime,
        exitTime,
        symbol,
        direction,
        lots,
        entryPrice,
        exitPrice,
        sl,
        tp,
        grossPnl,
        commission,
        barsHeld,
        closeReason,
        algoName,
        netPnl()   { return this.grossPnl - this.commission; },
        isWinner() { return this.netPnl() > 0; },
    };
}

// ── BTResult ──────────────────────────────────────────────────────────────────

export class BTResult {
    /**
     * @param {string}   symbol
     * @param {number}   timeframe
     * @param {number}   initialCapital
     * @param {object[]} trades        Array of BTTrade objects
     * @param {number[]} equityCurve   Equity value at each bar (from warmupBars onward)
     * @param {number}   durationSec   Wall-clock seconds the engine took
     */
    constructor(symbol, timeframe, initialCapital, trades, equityCurve, durationSec) {
        this.symbol         = symbol;
        this.timeframe      = timeframe;
        this.initialCapital = initialCapital;
        this.trades         = trades;
        this.equityCurve    = equityCurve;
        this.durationSec    = durationSec;
    }

    // ── Derived metrics ────────────────────────────────────────────────────────

    netProfit() {
        return this.trades.reduce((s, t) => s + t.netPnl(), 0);
    }

    finalCapital() {
        return this.initialCapital + this.netProfit();
    }

    returnPct() {
        return this.netProfit() / this.initialCapital * 100;
    }

    winRate() {
        if (this.trades.length === 0) return 0;
        const winners = this.trades.filter(t => t.isWinner()).length;
        return winners / this.trades.length;
    }

    profitFactor() {
        let grossWin = 0, grossLoss = 0;
        for (const t of this.trades) {
            const p = t.netPnl();
            if (p > 0) grossWin  += p;
            else        grossLoss += Math.abs(p);
        }
        return grossLoss < 1e-10 ? 999 : grossWin / grossLoss;
    }

    maxDrawdownPct() {
        const eq = this.equityCurve;
        if (eq.length === 0) return 0;
        let peak = eq[0];
        let maxDD = 0;
        for (const v of eq) {
            if (v > peak) peak = v;
            const dd = (peak - v) / peak;
            if (dd > maxDD) maxDD = dd;
        }
        return maxDD * 100;
    }

    sharpeRatio() {
        const eq = this.equityCurve;
        if (eq.length < 2) return 0;
        // bar-to-bar absolute returns
        const rets = [];
        for (let i = 1; i < eq.length; i++) {
            rets.push(eq[i] - eq[i - 1]);
        }
        const mean     = rets.reduce((s, r) => s + r, 0) / rets.length;
        const variance = rets.reduce((s, r) => s + (r - mean) ** 2, 0) / rets.length;
        const std      = Math.sqrt(variance);
        if (std < 1e-10) return 0;
        return (mean / std) * Math.sqrt(252);
    }

    sortinoRatio() {
        const eq = this.equityCurve;
        if (eq.length < 2) return 0;
        const rets = [];
        for (let i = 1; i < eq.length; i++) {
            rets.push(eq[i] - eq[i - 1]);
        }
        const mean     = rets.reduce((s, r) => s + r, 0) / rets.length;
        // downside deviation — only negative returns
        const negSqSum = rets.reduce((s, r) => s + (r < 0 ? r * r : 0), 0);
        const downStd  = Math.sqrt(negSqSum / rets.length);
        if (downStd < 1e-10) return 0;
        return (mean / downStd) * Math.sqrt(252);
    }

    calmarRatio() {
        const dd = this.maxDrawdownPct();
        if (dd < 1e-10) return 0;
        return this.returnPct() / dd;
    }

    avgWin() {
        const winners = this.trades.filter(t => t.isWinner());
        if (winners.length === 0) return 0;
        return winners.reduce((s, t) => s + t.netPnl(), 0) / winners.length;
    }

    avgLoss() {
        const losers = this.trades.filter(t => !t.isWinner());
        if (losers.length === 0) return 0;
        return losers.reduce((s, t) => s + t.netPnl(), 0) / losers.length;
    }

    expectancy() {
        const wr = this.winRate();
        return wr * this.avgWin() + (1 - wr) * this.avgLoss();
    }

    summary() {
        const pct = n => n.toFixed(2) + '%';
        const usd = n => '$' + n.toFixed(2);
        const f2  = n => n.toFixed(2);
        return [
            `Symbol        : ${this.symbol}`,
            `Timeframe     : ${this.timeframe}`,
            `Initial Cap   : ${usd(this.initialCapital)}`,
            `Final Cap     : ${usd(this.finalCapital())}`,
            `Net Profit    : ${usd(this.netProfit())}`,
            `Return        : ${pct(this.returnPct())}`,
            `Trades        : ${this.trades.length}`,
            `Win Rate      : ${pct(this.winRate() * 100)}`,
            `Profit Factor : ${f2(this.profitFactor())}`,
            `Max Drawdown  : ${pct(this.maxDrawdownPct())}`,
            `Sharpe Ratio  : ${f2(this.sharpeRatio())}`,
            `Sortino Ratio : ${f2(this.sortinoRatio())}`,
            `Calmar Ratio  : ${f2(this.calmarRatio())}`,
            `Avg Win       : ${usd(this.avgWin())}`,
            `Avg Loss      : ${usd(this.avgLoss())}`,
            `Expectancy    : ${usd(this.expectancy())}`,
            `Duration (s)  : ${this.durationSec.toFixed(3)}`,
        ].join('\n');
    }
}

// ── BacktestEngine ────────────────────────────────────────────────────────────

export class BacktestEngine {
    /**
     * @param {IAlgorithm} algo
     * @param {BTConfig|object} [cfg]  BTConfig instance or plain config object.
     */
    constructor(algo, cfg = new BTConfig()) {
        this._algo = algo;
        this._cfg  = cfg instanceof BTConfig ? cfg : new BTConfig(cfg);
    }

    /**
     * Run a backtest over the provided bar array.
     *
     * @param {object[]} bars    Bar-like objects: { timestamp, open, high, low, close, volume }
     * @param {string}   symbol
     * @param {number}   tf      Timeframe constant (seconds)
     * @returns {BTResult}
     */
    run(bars, symbol, tf) {
        const cfg      = this._cfg;
        const algo     = this._algo;
        const n        = bars.length;
        const startMs  = Date.now();
        const slippage = cfg.slippagePts * 0.00001;  // points → price units

        // Pre-compute the full-series ATR for SL/TP sizing
        const highs  = bars.map(b => b.high);
        const lows   = bars.map(b => b.low);
        const closes = bars.map(b => b.close);
        const atrArr = _atrFn(highs, lows, closes, 14);

        const trades      = [];
        const equityCurve = [];

        let capital   = cfg.initialCapital;
        let openTrade = null;   // active BTTrade or null

        for (let i = cfg.warmupBars; i < n; i++) {
            const bar  = bars[i];
            const next = (i + 1 < n) ? bars[i + 1] : null;

            // ── Step 1: Check SL/TP on current bar's high/low ─────────────────
            if (openTrade !== null) {
                let closed = false;

                if (openTrade.direction === 1 /* LONG */) {
                    if (bar.low <= openTrade.sl) {
                        // Stop-loss hit
                        openTrade.exitPrice   = openTrade.sl;
                        openTrade.exitTime    = bar.timestamp;
                        openTrade.barsHeld    = i - openTrade._entryBar;
                        openTrade.grossPnl    = (openTrade.sl - openTrade.entryPrice)
                                                * openTrade.lots * 100000;
                        openTrade.commission  = cfg.commissionUsd * openTrade.lots;
                        openTrade.closeReason = 'SL';
                        closed = true;
                    } else if (bar.high >= openTrade.tp) {
                        // Take-profit hit
                        openTrade.exitPrice   = openTrade.tp;
                        openTrade.exitTime    = bar.timestamp;
                        openTrade.barsHeld    = i - openTrade._entryBar;
                        openTrade.grossPnl    = (openTrade.tp - openTrade.entryPrice)
                                                * openTrade.lots * 100000;
                        openTrade.commission  = cfg.commissionUsd * openTrade.lots;
                        openTrade.closeReason = 'TP';
                        closed = true;
                    }
                } else /* SHORT */ {
                    if (bar.high >= openTrade.sl) {
                        // Stop-loss hit
                        openTrade.exitPrice   = openTrade.sl;
                        openTrade.exitTime    = bar.timestamp;
                        openTrade.barsHeld    = i - openTrade._entryBar;
                        openTrade.grossPnl    = (openTrade.entryPrice - openTrade.sl)
                                                * openTrade.lots * 100000;
                        openTrade.commission  = cfg.commissionUsd * openTrade.lots;
                        openTrade.closeReason = 'SL';
                        closed = true;
                    } else if (bar.low <= openTrade.tp) {
                        // Take-profit hit
                        openTrade.exitPrice   = openTrade.tp;
                        openTrade.exitTime    = bar.timestamp;
                        openTrade.barsHeld    = i - openTrade._entryBar;
                        openTrade.grossPnl    = (openTrade.entryPrice - openTrade.tp)
                                                * openTrade.lots * 100000;
                        openTrade.commission  = cfg.commissionUsd * openTrade.lots;
                        openTrade.closeReason = 'TP';
                        closed = true;
                    }
                }

                if (closed) {
                    capital += openTrade.netPnl();
                    trades.push(openTrade);
                    openTrade = null;
                }
            }

            // ── Step 2: No open trade + next bar exists → ask algo ────────────
            if (openTrade === null && next !== null) {
                const ind      = _buildIndicators(bars.slice(0, i + 1));
                const pat      = _buildPatterns(bars.slice(0, i + 1));
                const decision = algo.safeEvaluate(symbol, tf, bars, i + 1, ind, pat);

                if (decision.isActionable()) {
                    const atrVal = atrArr[i];
                    if (atrVal !== null && atrVal > 0) {
                        const isLong = decision.signal === AlgoSignal.BUY;

                        // Fill at next bar's open ± slippage
                        const fillPx = isLong
                            ? next.open + slippage
                            : next.open - slippage;

                        // ATR-based stop / target
                        const stopDist = atrVal * cfg.atrSlMult;
                        const sl = isLong
                            ? fillPx - stopDist
                            : fillPx + stopDist;
                        const tp = isLong
                            ? fillPx + stopDist * cfg.rrRatio
                            : fillPx - stopDist * cfg.rrRatio;

                        // Position sizing: risk fixed % of capital over stop distance
                        const _sizeMult = (algo && algo.metadata && algo.metadata.sizeMult != null)
                            ? algo.metadata.sizeMult : 1.0;
                        let lots = (capital * cfg.riskPct) / (stopDist * 100000) * _sizeMult;
                        lots = Math.round(lots * 100) / 100;           // round to 2 dp
                        lots = Math.max(0.01, Math.min(10.0, lots));   // clamp [0.01, 10]

                        openTrade = makeBTTrade({
                            entryTime : next.timestamp,
                            symbol,
                            direction : isLong ? 1 : -1,
                            lots,
                            entryPrice: fillPx,
                            sl,
                            tp,
                            algoName  : algo.name(),
                        });
                        // Internal marker for barsHeld computation
                        openTrade._entryBar = i + 1;
                    }
                }
            }

            // ── Step 3: Record equity curve ───────────────────────────────────
            let floatingPnl = 0;
            if (openTrade !== null) {
                floatingPnl = openTrade.direction === 1
                    ? (bar.close - openTrade.entryPrice) * openTrade.lots * 100000
                    : (openTrade.entryPrice - bar.close) * openTrade.lots * 100000;
            }
            equityCurve.push(capital + floatingPnl);
        }

        // ── Step 4: Force-close any open trade at last bar's close ────────────
        if (openTrade !== null && n > 0) {
            const lastBar           = bars[n - 1];
            openTrade.exitPrice     = lastBar.close;
            openTrade.exitTime      = lastBar.timestamp;
            openTrade.barsHeld      = (n - 1) - openTrade._entryBar;
            openTrade.grossPnl      = openTrade.direction === 1
                ? (lastBar.close - openTrade.entryPrice) * openTrade.lots * 100000
                : (openTrade.entryPrice - lastBar.close) * openTrade.lots * 100000;
            openTrade.commission    = cfg.commissionUsd * openTrade.lots;
            openTrade.closeReason   = 'END';
            capital += openTrade.netPnl();
            trades.push(openTrade);
        }

        const durationSec = (Date.now() - startMs) / 1000;
        return new BTResult(symbol, tf, cfg.initialCapital, trades, equityCurve, durationSec);
    }
}
